pkgs/by-name/pa/pandoc/package.nix

```diff
@@ -1,8 +1,8 @@
-{ stdenv, lib, haskellPackages, haskell, removeReferencesTo, installShellFiles }:
+{ stdenv, lib, haskellPackages, haskell, removeReferencesTo, installShellFiles, pandoc-cli ? haskellPackages.pandoc-cli }:
```

Comment:
In other words, we can't be confident that the input `haskellPackages` and `haskellPackages.pandoc-cli` are in sync so we should pass in a single `haskellPackages` and pull from that?

Would this work?

```patch
diff --git a/pkgs/by-name/pa/pandoc/package.nix b/pkgs/by-name/pa/pandoc/package.nix
index b956ee51bb02..b1bdd88f2b37 100644
--- a/pkgs/by-name/pa/pandoc/package.nix
+++ b/pkgs/by-name/pa/pandoc/package.nix
@@ -1,7 +1,8 @@
-{ stdenv, lib, haskellPackages, haskell, removeReferencesTo, installShellFiles, pandoc-cli ? haskellPackages.pandoc-cli }:
+{ stdenv, lib, haskellPackages, haskell, removeReferencesTo, installShellFiles, pandocVersion ? "pandoc-cli" }:

 let
   # Since pandoc 3.0 the pandoc binary resides in the pandoc-cli package.
+  pandoc-cli = lib.getAttr pandocVersion haskellPackages;
   static = haskell.lib.compose.justStaticExecutables pandoc-cli;
```

Reply:
Very stringly. I'd go with (p: p.pandoc-cli) as default argument.

Or simply override the haskellPackages set to have the right pandoc-cli when passing in. I think I'd favor that.

Reply:
But it is the cleaner abstraction, imo. The package wouldn't need to know about the overriding at all.

I am fine either way but I'd love to see a benchmark before throwing the idea away.

Reply:
Why don’t we just replace all `haskellPackages` references with `pandoc-cli.scope`?

Reply:
Well that would only make sense to eliminate the haskellPackages input completely, right?

The problem with that approach is that the default callPackage for by-name would not pick up that pandoc-cli is supposed to be haskellPackages.pandoc-cli. So I don’t think it’s super helpful, is it?

Reply:
True. Then I guess the current state of the PR is basically fine?

Reply:
> But it is the cleaner abstraction, imo. The package wouldn't need to know about the overriding at all.
> 
> I am fine either way but I'd love to see a benchmark before throwing the idea away.

I'm happy with whatever route we take here and don't want to derail a decision, but for future reference I'm curious about the comment @maralorn made on benchmarks. Would time spent running `nix eval` or `nix build` (with caching) be enough here? If not would you have a recommendation for how you'd go about doing this?

For example, I can run something like the following but I'm not sure how to evaluate if the results here are meaningless.

```bash
# Three warmups before running each command 30 times.
> hyperfine --warmup 3 --min-runs 30 \
    "git checkout quarto-pandoc && nix build .#pandoc_3_5" \
    "git checkout quarto-pandoc-2 && nix build .#pandoc_3_5"

Benchmark 1: git checkout quarto-pandoc && nix build .#pandoc_3_5
  Time (mean ± σ):      1.107 s ±  0.023 s    [User: 0.567 s, System: 0.593 s]
  Range (min … max):    1.074 s …  1.150 s    30 runs
 
Benchmark 2: git checkout quarto-pandoc-2 && nix build .#pandoc_3_5
  Time (mean ± σ):      1.134 s ±  0.012 s    [User: 0.582 s, System: 0.601 s]
  Range (min … max):    1.115 s …  1.158 s    30 runs
 
Summary
  git checkout quarto-pandoc && nix build .#pandoc_3_5 ran
    1.02 ± 0.02 times faster than git checkout quarto-pandoc-2 && nix build .#pandoc_3_5
```

Reply:
I have no clue whether that is a robust way of benchmarking. But a) I would use just nix .#eval and b) I would be afraid of this running into eval caching.

Reply:
Agreed, I'd bypass the flake since it may cache evals (i.e. just use `nix-instantiate` or `nix-build`). But the basic idea seems right. I suspect the difference is within the error of the test, though, so it is hard to conclusively say one or the other is better (this is in my experience almost always the case with such small optimizations in Nix code).